File: algorithm/go/top_k.go
package algorithm

func BotK(nums []int, l, r, k int) int {
	if l == r {
		return nums[l]
	}
	i, j, mid := l-1, r+1, nums[(l+r)/2]
	for i < j {
		for i++; nums[i] < mid; i++ {
		}
		for j--; nums[j] > mid; j-- {
		}
		if i < j {
			nums[i], nums[j] = nums[j], nums[i]
		}
	}
	if j-l+1 >= k {
		return BotK(nums, l, j, k)
	}
	return BotK(nums, j+1, r, k-(j-l+1))
}

// 从 1 开始
func TopK(nums []int, l, r, k int) int {
	return BotK(nums, l, r, r-l+1-k+1)
}

File: algorithm/cpp/top_k.cpp
#include <iostream>

using namespace std;

const int N = 100010;
int a[N];

int TopK(int l, int r, int k)
{
    if (l == r)
        return a[l];
    int i = l - 1, j = r + 1, mid = a[l + r >> 1];
    while (i < j)
    {
        do
            i++;
        while (a[i] < mid);
        do
            j--;
        while (a[j] > mid);
        if (i < j)
            swap(a[i], a[j]);
    }
    if (j - l + 1 >= k)
        return TopK(l, j, k);
    return TopK(j + 1, r, k - (j - l + 1));
}

int main()
{
    int n, k;
    cin >> n >> k;
    for (int i = 0; i < n; i++)
        cin >> a[i];
    cout << TopK(0, n - 1, k);
    return 0;
}